cmd/src/batch_exec.go

```diff
@@ -182,75 +155,65 @@ func executeBatchSpecInWorkspaces(ctx context.Context, flags *executorModeFlags)
-
-	if len(tasks) != 1 {
-		return errors.New("invalid input, didn't yield exactly one task")
-	}
-
-	task := tasks[0]
+	task := convertWorkspace(input)
 
 	if len(task.Steps) == 0 {
 		return errors.New("invalid execution, no steps to process")
 	}
 
-	{
-		ui.PreparingContainerImages()
-		_, err = svc.EnsureDockerImages(
-			ctx,
-			task.Steps,
-			execPullParallelism,
-			ui.PreparingContainerImagesProgress,
-		)
-		if err != nil {
-			return err
-		}
-		ui.PreparingContainerImagesSuccess()
-	}
-
-	coord := svc.NewCoordinator(
-		repozip.NewNoopRegistry(),
-		log.NewNoopManager(),
-		executor.NewCoordinatorOpts{
-			Creator:     workspace.NewExecutorWorkspaceCreator(tempDir, repoDir),
-			Cache:       &executor.ServerSideCache{CacheDir: flags.cacheDir, Writer: ui},
-			Parallelism: 1,
-			// TODO: Should be slightly less than the executor timeout. Can we somehow read that?
-			Timeout: flags.timeout,
-			TempDir: tempDir,
-			// Don't allow to read from env.
-			GlobalEnv: []string{},
-			// Temporarily prevent the ability to sending a batch spec with a mount for server-side processing.
-			IsRemote: true,
-		},
-	)
+	imageCache := docker.NewImageCache()
 
-	// `src batch exec` uses server-side caching for changeset specs, so we
-	// only need to call `CheckStepResultsCache` to make sure that per-step cache entries
-	// are loaded and set on the tasks.
-	if err := coord.CheckStepResultsCache(
+	ui.PreparingContainerImages()
+	_, err = service.New(&service.Opts{}).EnsureDockerImages(
 		ctx,
-		tasks,
-		// Don't expose the executor env, we don't allow env forwarding anyways.
-		[]string{},
-	); err != nil {
+		imageCache,
+		task.Steps,
+		execPullParallelism,
+		ui.PreparingContainerImagesProgress,
+	)
+	if err != nil {
 		return err
 	}
+	ui.PreparingContainerImagesSuccess()
+
+	// Empty for now until we support secrets or env var settings in SSBC.
+	globalEnv := []string{}
+	isRemote := true
 
-	// These arguments are unused in the json logs implementation, but the interface
-	// dictates them.
+	// Set up the execution UI.
 	taskExecUI := ui.ExecutingTasks(false, 1)
-	err = coord.Execute(ctx, tasks, taskExecUI)
-	if err != nil {
-		taskExecUI.Failed(err)
-		return err
+	taskExecUI.Start([]*executor.Task{task})
+	taskExecUI.TaskStarted(task)
+
+	opts := &executor.RunStepsOpts{
+		Logger:      &log.NoopTaskLogger{},
+		WC:          workspace.NewExecutorWorkspaceCreator(tempDir, repoDir),
+		EnsureImage: imageCache.Ensure,
+		Task:        task,
+		// TODO: Should be slightly less than the executor timeout. Can we somehow read that?
+		Timeout:   flags.timeout,
+		TempDir:   tempDir,
+		GlobalEnv: globalEnv,
+		// Temporarily prevent the ability to sending a batch spec with a mount for server-side processing.
+		IsRemote:    isRemote,
+		RepoArchive: &repozip.NoopArchive{},
+		UI:          taskExecUI.StepsExecutionUI(task),
 	}
+	results, err := executor.RunSteps(ctx, opts)
```

Comment:
Running steps directly, no coordinator nor executor needed here! They're only concurrency manager wrappers that are overhead in SSBC and made it hard to understand which flags are actually required and if we might be hitting sg, hence the deadClient before.